src/zenml/integrations/hyperai/service_connectors/hyperai_service_connector.py

```diff
@@ -0,0 +1,386 @@
+            logger.error("Bad host key: %s", e)
+        except paramiko.ssh_exception.AuthenticationException as e:
+            logger.error("Authentication failed: %s", e)
+        except paramiko.ssh_exception.SSHException as e:
+            logger.error(
+                "SSH error: %s. A common cause for this error is selection of the wrong key type in your service connector.",
+                e,
+            )
+        except Exception as e:
+            logger.error(
+                "Unknown error while connecting to HyperAI instance: %s. Please check your network connection, IP address, and authentication details.",
+                e,
+            )
+
+        return None
+
+    def _authorize_client(self) -> None:
+        """Verify that the client can authenticate with the HyperAI instance.
+
+        Raises:
+            BadHostKeyException: If the host key is invalid.
+            AuthenticationException: If the authentication credentials are
+                invalid.
+            SSHException: If there is an SSH related error.
+        """
+        logger.info("Verifying connection to HyperAI instance...")
+
+        try:
+            paramiko_client = self._create_paramiko_client()
+            if paramiko_client is None:
+                raise RuntimeError(
+                    "Could not authorize client due to missing paramiko client."
+                )
+            else:
+                paramiko_client.close()
+        except Exception as e:
+            logger.error(
+                "Unknown error while connecting to HyperAI instance: %s. Please check your network connection, IP address, and authentication details.",
+                e,
+            )
+
+    def _connect_to_resource(
+        self,
+        **kwargs: Any,
+    ) -> Any:
+        """Connect to a HyperAI instance. Returns an authenticated SSH client.
+
+        Args:
+            kwargs: Additional implementation specific keyword arguments to pass
+                to the session or client constructor.
+
+        Returns:
+            An authenticated Paramiko SSH client.
+        """
+        logger.info("Connecting to HyperAI instance...")
+        assert self.resource_id is not None
+
+        try:
+            paramiko_client = self._create_paramiko_client()
+            if paramiko_client is None:
+                raise RuntimeError(
+                    "Could not connect to instance due to missing paramiko client."
+                )
+            else:
+                return paramiko_client
+        except Exception as e:
+            logger.error(
+                "Unknown error while connecting to HyperAI instance: %s. Please check your network connection, IP address, and authentication details.",
+                e,
+            )
+
+    def _configure_local_client(
+        self,
+        **kwargs: Any,
+    ) -> None:
+        """There is no local client for the HyperAI connector, so it does nothing."""
+        pass
```

Comment:
This sort of thing might be useful when someone needs to access the HyperAI instance to debug something. You could use the stored SSH key / passphrase to configure the local SSH client.

If you choose not to implement this method though, could you please raise `NotImplementedError` instead ? This is better handled in the CLI.

Reply:
If I would want to implement it, I could imagine that I would return the `paramiko.client.SSHClient` by calling `self._create_paramiko_client()` - or does this def have a different purpose?

Reply:
This method serves a different purpose. It's basically meant to be the exact opposite of discovering and lifting up credentials from your local machine. In this case, it would most likely create a file in the `~/.ssh/` folder containing the  key stored in the connector configuration and instruct the user on how to use e.g. by showing some example `ssh -i <private-key-file> <hostname>` commands.

Reply:
Hm, I'd say we leave that out for now. I'll make sure to raise `NotImplementedError`.

Reply:
https://github.com/zenml-io/zenml/pull/2372/commits/6ca8b97a0e257e87aa427dd717ffd2419870d4dd